Shared/IISFunctions/Get-IISModules.ps1

```diff
@@ -0,0 +1,159 @@
+﻿# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+function Get-IISModules {
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory = $true)]
+        [System.Xml.XmlNode]$ApplicationHostConfig,
+
+        [Parameter(Mandatory = $false)]
+        [scriptblock]$CatchActionFunction
+    )
+
+    begin {
+        Write-Verbose "Calling: $($MyInvocation.MyCommand)"
+        function ConvertCmdEnvironmentVariableToFullPath {
+            [CmdletBinding()]
+            [OutputType([System.String])]
+            param(
+                [String]$PathWithVar
+            )
+
+            $returnPath = $null
+            if (-not([String]::IsNullOrEmpty($PathWithVar))) {
+                # Assuming that we have the env var always at the beginning of the string and no other vars within the string
+                # Example: %windir%\system32\someexample.dll
+                $preparedPath = ($PathWithVar.Split("%", [System.StringSplitOptions]::RemoveEmptyEntries))
+                if ($preparedPath.Count -eq 2) {
+                    if ($preparedPath[0] -notmatch "\\.+\\") {
+                        $varPath = [System.Environment]::GetEnvironmentVariable($preparedPath[0])
+                        $returnPath = [String]::Join("", $varPath, $($preparedPath[1]))
+                    }
+                }
+            }
+            return $returnPath
+        }
+
+        function GetIISModulesLoaded {
+            [CmdletBinding()]
+            param(
+                [Parameter(Mandatory = $true)]
+                [System.Xml.XmlNode]$Xml
+            )
+            process {
+                try {
+                    $iisModulesList = New-Object 'System.Collections.Generic.List[object]'
+                    $globalModules = $Xml.configuration.'system.webServer'.globalModules.add
+                    if ($globalModules.Count -ge 1) {
+                        Write-Verbose "At least one module is loaded by IIS"
+                        foreach ($m in $globalModules) {
+                            Write-Verbose "Now processing module: $($m.Name)"
+                            $mInfoObject = [PSCustomObject]@{
+                                Name             = $m.Name
+                                Path             = $null
+                                Signed           = $false
+                                SignatureDetails = $null
+                            }
+
+                            $moduleFilePath = $m.Image
+                            if ($moduleFilePath -match "\%.+\%") {
+                                # Overwrite path when environment variables were found
+                                Write-Verbose "Environment variable found in path: $moduleFilePath"
+                                $moduleFilePath = ConvertCmdEnvironmentVariableToFullPath -PathWithVar $m.Image
+                            }
```

Comment:
To make this easier to read, maybe have this be like this.
```
$moduleFilePath = GetModulePath $m.Image 

Function GetModulePath {
param($path) 
if ($path -match "\%.+\%") { #code in convertCmdEnvironmentVariableToFullPath}
return $path
}
```

Reply:
Good idea! Code adjusted.